nginx/modsecurity/ngx_http_modsecurity.c

```diff
@@ -975,267 +991,170 @@ ngx_http_modsecurity_init_process(ngx_cycle_t *cycle)
-
-    if (ngx_http_set_pool_ctx(r, ctx, ngx_http_modsecurity) != NGX_OK) {
-        return NGX_ERROR;
-    }
-
-    /* load request to request rec */
+    // Load request to request rec
     if (ngx_http_modsecurity_load_request(r) != NGX_OK
         || ngx_http_modsecurity_load_headers_in(r) != NGX_OK) {
 
-        return NGX_HTTP_INTERNAL_SERVER_ERROR;
+        mod_ctx->status_code = NGX_HTTP_INTERNAL_SERVER_ERROR;
+        return;
     }
 
-    /* processing request headers */
-    rc = ngx_http_modsecurity_status(r, modsecProcessRequestHeaders(ctx->req));
+    // Processing request headers
+    ngx_int_t rc = ngx_http_modsecurity_status(r, modsecProcessRequestHeaders(mod_ctx->req));
 
     if (rc != NGX_DECLINED) {
-        return rc;
+        mod_ctx->status_code = rc;
+        return;
     }
 
-    if (modsecContextState(ctx->req) == MODSEC_DISABLED) {
-        return NGX_DECLINED;
+    if (modsecContextState(mod_ctx->req) == MODSEC_DISABLED) {
+        mod_ctx->status_code = NGX_DECLINED;
+        return;
     }
 
-
-    if (modsecIsRequestBodyAccessEnabled(ctx->req) && (r->headers_in.content_length || r->headers_in.chunked)) {
-        rc = ngx_http_read_client_request_body(r, ngx_http_modsecurity_body_handler);
-        if (rc >= NGX_HTTP_SPECIAL_RESPONSE) {
-            return rc;
+    if (modsecIsRequestBodyAccessEnabled(mod_ctx->req) && (r->headers_in.content_length || r->headers_in.chunked)) {
+        if (ngx_http_modsecurity_load_request_body(r) != NGX_OK) {
+            mod_ctx->status_code = NGX_HTTP_INTERNAL_SERVER_ERROR;
+            return;
         }
-
-        return NGX_DONE;
     }
 
-    return ngx_http_modsecurity_status(r, modsecProcessRequestBody(ctx->req));
+    // The name of modsecProcessRequestBody is a bit misleading. This function call is needed even to just process GET args.
+    rc = ngx_http_modsecurity_status(r, modsecProcessRequestBody(mod_ctx->req));
+    mod_ctx->status_code = rc;
 }
 
 
 static void
-ngx_http_modsecurity_body_handler(ngx_http_request_t *r)
+ngx_http_modsecurity_thread_completion(ngx_event_t *ev)
 {
-    ngx_http_modsecurity_ctx_t    *ctx;
-    ngx_int_t                      rc;
-
-    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, r->connection->log, 0, "modSecurity: body handler");
-
-    ctx = ngx_http_get_module_ctx(r, ngx_http_modsecurity);
-
-    if (ngx_http_modsecurity_load_request_body(r) != NGX_OK) {
-        return ngx_http_finalize_request(r, NGX_HTTP_INTERNAL_SERVER_ERROR);
-    }
-
-    rc = ngx_http_modsecurity_status(r, modsecProcessRequestBody(ctx->req));
+    // executed in nginx event loop after thread task is done, in order to pick up and continue processing request
 
-    if (rc != NGX_DECLINED) {
-        return ngx_http_finalize_request(r, rc);
-    }
+    ngx_http_request_t  *r;
 
-    if (ngx_http_modsecurity_save_request_body(r) != NGX_OK
-            || ngx_http_modsecurity_save_headers_in(r) != NGX_OK ) {
+    r = ev->data;
+    r->main->blocked--;
```

Comment:
Is blocked getting used anywhere?

Reply:
This is a sore spot you found. I also don't really know. I copied this from other plugins that does things async. Also tried without having this, and it makes nginx hang, so it must be important.

Reply:
OK, better to keep it around.